### ISSUE TITLE
Some ideas...

### ISSUE THREAD

**Issue by Tony-X** [nikomatsakis/moro#2]
You asked for feedbacks in Learn Some Rust this Month so I'm here... 
The buggy code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=66455d3e61217d86ec4839c926619bf0

I like the scope approach since it allows the programmer to effectively construct a single eventloop for different tasks. Though I think there is a fair question to ask -- how do we teach the programmers to put the sender and receiver code in the same scope?

Here is another way to solve this type of problems. I tried to generalize but it might still be too specific to this issue.

First, In many functional programming languages there is a common combinator (in Scala it's called [sequence](https://www.scala-lang.org/api/2.13.5/scala/concurrent/Future$.html#sequence[A,CC[X]%3C:IterableOnce[X],To](in:CC[scala.concurrent.Future[A]])(implicitbf:scala.collection.BuildFrom[CC[scala.concurrent.Future[A]],A,To],implicitexecutor:scala.concurrent.ExecutionContext):scala.concurrent.Future[To])) that does `[Future<Output=T>] -> Future<Output=[T]>` or the iterator version `Iterator<Item=Future<Output=T>> -> Future<Output=Iterator<Item=T>>` . This shares some sprit with `FutureUnordered` but it is functional as it doesn't work with a mutable construct and provide a nice abstraction. 

We can build a Rust version of sequence and provide another abstraction which is the looped [select](https://docs.rs/tokio/0.2.11/tokio/macro.select.html), lets call it `looped_select!` (i'm sure we can find better names!). Then the code becomes

```rust
    let mut receivers =  vec![];
    let mut senders = vec![];
    for host in 0..replicas {
        let (sender, receiver) = channel(2);
        senders .push(sender);
        receivers .push(replica(host, receiver));
    }
   let senders = sequence(senders);
   let receivers = sequence(senders);
   // some pinning 
   looped_select! {
     recv_stats_iter = receivers = { // consume receiver stats } 
     _ = senders { // noop }
   }
```

The fact that the granularity of the results is all stats is probably less desirable but should be fine. If it is really an issue, we need to put more thoughts in offering better polling for multiple Streams, i.e. Iterators of Futures, as I believe that alone will solve problems of such kind. Still, we need to teach people this pattern like "make sure you pull all iterators of futures together as they may block on each other"

**Comment by nikomatsakis** [comment]
@Tony-X 

Can you say a bit more about this:

>  how do we teach the programmers to put the sender and receiver code in the same scope?

I'm not sure if I understand what you're referring to here, but I don't think that senders and receivers have to be in the same scope necessarily. The moro API guarantees that the "scope task" plus every spawned task will execute concurrently with one another and hence make progress, regardless of what those subtasks do (well, as long as they await periodically). Since tokio guarantees that the various *tasks* themselves make progress, this implies that the channels will eventually be drained, and no deadlock will occur, regardless of whether the sender/receiver are in the same scope or not.

Regarding the `looped_select!` construct, I'm not sure what it's trying to achieve *exactly* but it seems rather less obvious than moro. For example, if I had `looped_select!` just over the receivers, but not the senders, or vice versa, than things would deadlock, right? 

To put it another way, the `looped_select!` construct looks like "do these things concurrently". If you follow a pattern where every task (a) prepares some work and then (b) terminates in a `looped_select!`, things should work fine, but if you want to intersperse spawning of tasks and awaiting things, and then finally running everything to completion, it won't work so well. Something like this:

```rust
moro::async_scope!(|scope| {
    while let Some(item) = read_network_channel().await {
        scope.spawn(process_incoming_item(item));
    }
}).infallible().await;
```

**Comment by Tony-X** [comment]
Thanks for sharing your thoughts. The point regarding `looped_select!` is insightful. I agree that it can't deal with dynamically create futures. 

> how do we teach the programmers to put the sender and receiver code in the same scope?

I meant something like the following which will hang since receivers will block. 

```rust
    let mut senders = vec![];
    let mut receivers = vec![];

    for host in 0..replicas {
        let (sender, receiver) = channel(2);
        senders.push(sender);
        receivers.push(receiver));
    }

    moro::async_scope!(|scope| {
        // setup receiver side futures and await them
    }).infallible().await;
    
    // await senders

```

However, when I tried to implement this the compiler stopped me since I can't move the receivers into the scope (a good thing!). 

I'm convinced that it is best to just `spawn` and let the runtime handles theses inter-dependent futures' progress.
